File: hackerrank/problem solving/algorithms/2 - implementation/17 - picking numbers/with morfo library/picking_numbers.cs
// https://www.hackerrank.com/challenges/picking-numbers/problem?isFullScreen=true

using Morfo.Collections;
using Morfo.IO;

Reader.SkipInputLn();
var freqMap = Counter.Of(Reader.ReadLn<int, int[]>());
Console.WriteLine(PickingNumbers(freqMap));

// k: length of freqMap
// 1 <= k <= 99
// T: O(k) = O(99) = O(1)
// S: O(1) extra space
int PickingNumbers(Dictionary<int, int> freqMap)
{
    int maxLen = 0;
    foreach (int num in freqMap.Keys)
    {
        int current = freqMap[num] + (freqMap.ContainsKey(num + 1) ? freqMap[num + 1] : 0);
        maxLen = Math.Max(maxLen, current);
    }
    return maxLen;
}

File: hackerrank/problem solving/algorithms/2 - implementation/17 - picking numbers/with morfo library/picking_numbers.ts
// https://www.hackerrank.com/challenges/picking-numbers/problem?isFullScreen=true

import { counter, readLn, skipInputLn } from 'morfo';

function main() {
  skipInputLn();
  const freqMap: Map<number, number> = counter(readLn(Number));
  console.log(pickingNumbers(freqMap));
}

// k: length of freqMap
// 1 <= k <= 99
// T: O(k) = O(99) = O(1)
// S: O(1) extra space
function pickingNumbers(freqMap: Map<number, number>): number {
  let maxLen = 0;
  for (const [num, _] of freqMap) {
    const current: number = freqMap.get(num)! + (freqMap.get(num + 1) || 0);
    maxLen = Math.max(maxLen, current);
  }
  return maxLen;
}

main();
